Calls both processes in a single file

p-galligan opened the pull request:
Not sure this is the best approach, but it does give the ability to wrap up both files in a single python call using `main.py` with necessary arguments. Works best if each directory is different.

Commit:
call from single file

calls processes from single file

Commit:
updates

updates imported libraries

Commit:
add pycache to ignore

ignores pycache

Commit:
remove pycache

removes existing pycache files

Commit:
remove log

removes derivative log

helrond commented:
OK, I took a look at this and a couple of comments:

My understanding is that we'll be working from a data structure that looks like the following (where 12345678910 and 12345678911 are AS ref_ids):
```
12345678910/
    123456789_001.tiff
    123456789_002.tiff
    ...
12345678911/
    12345678911_001.tiff
    12345678911_002.tiff
    ....
```
Based on what I'm seeing here, you'd have to point this script at the 12345678910 and 12345678911 directories individually, which means you'd have to know the ref_id before you ran this script, and also that you'd have to run it individually on each ref_id directory. I think we need to make this run one level up, so that it loops through the directories and processes each one in turn. That means your main wrapper class should store the id of the directory it's working on, and can pass that to both the derivative creation and manifest creation processes.

Second, there's a lot of setting of attributes on a class, and then passing those same attributes into another class. I think you want to take a look at this so you're not passing tons of arguments around - it makes understanding the logic much harder to follow. For example, `source_dir`, `skip`, and `default_options` get set as attributes on RunProcesses, but are then just passed to the init method of DerivativeMaker, and are not used in ManifestMaker at all. I would suggest trying to isolate variables and attributes to the places where they're needed. In many cases, you can do that simply by passing arguments to the method you're going to eventually call (usually `run`) rather than the class itself. 

Last, I'm wondering if it would make sense to pull the upload to S3 out as a separate process. It seems like that could fail in multiple ways and/or you might want to have some conditional logic which would allow you to not upload or upload assets that have already been created. I think it would also help clarify the different functions and simplify what you have to pass to each.

There are some other minor things but I think it makes more sense to look at that in code review.

Commit:
update for readability

Updates to move script one level higher, and separate out as much as possible.

p-galligan commented:
@helrond Ok, what I've done is move everything one level higher in the `main` wrapper. It now only takes a two arguments (source_dir and skip), and generates single derivative and manifest directories in that source dir where all derivative and manifest files will get written to.

I've also separated out the aws upload and put it in the main file (I'm happy to put that elsewhere in its own file though). Right now it checks if a file already exists in AWS and only uploads it if it doesn't. There's probably some improvements to be made, but it is functional.

I hope I've managed to hit your point about arguments in class inits vs. methods. I've tried to separate those out as much as possible.

helrond commented:
Yeah this looks MUCH better! Now doing something different with the IDs is much easier.

I think I would argue for putting the s3 stuff in a separate file - it helps keep that main file clean and readable.

Last, you're doing a lot of path manipulation with format strings, but you should absolutely use os.path.join instead. It will avoid the trailing slash issues.

Commit:
Separate AWS

Separate AWS actions from main file. Changes string injection in certain places to `os.path.join` when it makes sense.

Commit:
Updating docstrings

Updates docstrings for updated code.

Commit:
Update to include shortuuid

Updates to include shortuuid implementation. This includes breaking out files even more, and updaes documentation.

p-galligan commented:
Ok @helrond , I've been working on the assumption that we will have a directory of directories as outlined above. And each directory will be an AS refid.

With that process, I've implemented shortuuid to work. I've broken out the get request against ArchivesSpace to its own separate file, which runs before derivative and manifest creation. If it can find a matching archival object, it gets the title and date, and passes that to the manifest maker. Otherwise, it passes on making derivatives and manifests. 

That way we can get the AO id, and then just make the derivatives with the new shortuuid, and the manifest maker can just work off of those files. I think that reduces duplication of work and is probably the easiest way to go about it. Only downside is that there's nothing tying the derivatives and the manifest _back_ to the original refid right now.

helrond commented on main.py:
```diff
@@ -0,0 +1,56 @@
+        for directory in directories:
+            identifier = directory.split('/')[-1]
+            ao, title, date = archival_object.run(identifier)
+            uuid = shortuuid.uuid()
```
```suggestion
            uuid = shortuuid.uuid(name=ao["uri"])
```
See https://github.com/RockefellerArchiveCenter/pisces/blob/development/fetcher/helpers.py#L73

p-galligan replied:
Update. I moved this generation down into the conditional to only make the shortuuid if there's a matching AO. Otherwise it wouldn't be able to get `ao["uri"] from an empty dict which is produced if there's no match.

helrond commented on create_derivatives.py:
```diff
@@ -47,27 +43,28 @@ def run(self):
+        original_file = os.path.join(start_directory, file)
+        new_id = file.replace(file.split("_")[0], uuid)
+        fname = new_id.split(".")[0]
+        derivative_file = "{}.jp2".format(os.path.join(end_directory, fname))
```
```suggestion
        derivative_file = os.path.join(end_directory, "{}.jp2".format(fname))
```

helrond reviewed (changes requested):
A couple of minor tweaks. I do have some other questions but I think those are better saved for code review.

p-galligan committed:
Update create_derivatives.py

Co-authored-by: Hillel Arnold <helrond@hotmail.com>

Commit:
update get_ao

Updates archival object getting so it consistently returns a dictionary.

Move shortuuid to conditional. Only makes one if archival object is found.

Commit:
add pdf generation

Adds pdf creation script.

Commit:
pdf testing

tests pdf creation

Commit:
Update manifest and pdf

Updates manifest and pdf generation. Manifests now include service sections.

Commit:
update readme

Updates aws_upload section of readme to include s3 path where images are pushed.
